core/src/actors/mining_manager/handlers.rs

```diff
@@ -0,0 +1,100 @@
+use actix::{ActorFuture, Context, ContextFutureSpawner, Handler, System, WrapFuture};
+
+use super::MiningManager;
+use crate::actors::chain_manager::messages::GetHighestCheckpointBeacon;
+use crate::actors::chain_manager::ChainManager;
+use crate::actors::epoch_manager::messages::EpochNotification;
+
+use witnet_data_structures::chain::{
+    Hash, Hashable, LeadershipProof, Secp256k1Signature, Signature,
+};
+
+use crate::actors::chain_manager::messages::BuildBlock;
+use crate::actors::reputation_manager::messages::ValidatePoE;
+use crate::actors::reputation_manager::ReputationManager;
+use log::{debug, error};
+
+////////////////////////////////////////////////////////////////////////////////////////
+// ACTOR MESSAGE HANDLERS
+////////////////////////////////////////////////////////////////////////////////////////
+
+/// Payload for the notification for all epochs
+#[derive(Clone, Debug)]
+pub struct EveryEpochPayload;
+
+/// Handler for EpochNotification<EveryEpochPayload>
+impl Handler<EpochNotification<EveryEpochPayload>> for MiningManager {
+    type Result = ();
+
+    fn handle(&mut self, msg: EpochNotification<EveryEpochPayload>, ctx: &mut Context<Self>) {
+        debug!("Periodic epoch notification received {:?}", msg.checkpoint);
+
+        let chain_manager_addr = System::current().registry().get::<ChainManager>();
+        chain_manager_addr
+            .send(GetHighestCheckpointBeacon)
+            .into_actor(self)
+            .then(move |beacon_msg, act, ctx| {
+                // Check eligibility
+                // S(H(beacon))
+                let mut beacon = match beacon_msg {
+                    Ok(Ok(b)) => b,
+                    _ => return actix::fut::err(()),
+                };
+                if beacon.checkpoint > msg.checkpoint {
+                    // We got a block from the future
+                    error!("The current highest checkpoint beacon is from the future");
+                    return actix::fut::err(());
+                }
+                if beacon.checkpoint == msg.checkpoint {
+                    // For some reason we already got a valid block for this epoch
+                    // TODO: Check eligibility anyway?
+                }
+                // TODO: this is correct, right?
+                beacon.checkpoint = msg.checkpoint;
```

Comment:
This is right as long as we update the block digest of the `HighestCheckpointBeacon` in `ChainManager` at some point.